### ISSUE TITLE
Change QNEthernet's setDnsServerIP() to setDNSServerIP()

### ISSUE THREAD

**Issue by ssilverman** [JAndrassy/Arduino-Networking-API#8]
https://github.com/JAndrassy/Arduino-Networking-API/blob/639d731a124b7f7cdd5c70a7d48fe6c3b60956aa/ArduinoNetAPILibs.md?plain=1#L112

**Comment by JAndrassy** [comment]
I don't understand. Ethernet libraries have `setDnsServerIP` and alternative is `setDNS` as in WiFi libraries

**Comment by ssilverman** [comment]
Before I can answer that, let me ask you this question: Are you trying to follow Arduino's Ethernet API ([here](https://www.arduino.cc/reference/en/libraries/ethernet/)) or the WiFiNINI API ([here](https://www.arduino.cc/reference/en/libraries/wifinina/))?

There's a case where the Ethernet API uses [MACAddress()](https://www.arduino.cc/reference/en/libraries/ethernet/ethernet.macaddress/), while the WiFiNINA API uses [macAddress()](https://www.arduino.cc/reference/en/libraries/wifinina/wifi.macaddress/). Your project uses "macAddress()" and not "MACAddress()" for the common spelling. I agree with this because it keeps the naming style consistent. i.e. "start with a small-case letter". I tried to keep my API consistent with this naming scheme as well.

Look also at [setMACAddress()](https://www.arduino.cc/reference/en/libraries/ethernet/ethernet.setmacaddress/) vs. [setDnsServerIP()](https://www.arduino.cc/reference/en/libraries/ethernet/ethernet.setdnsserverip/) in the Ethernet API. Those are inconsistent. In the same way that you've chosen consistency with "macAddress()" vs. "MACAddress()", I've also chosen to use "setDNSServerIP()" vs. "setDnsServerIP()".

Now, to keep my own API internally consistent, I have "setDNSServerIP()" and "dnsServerIP()" for both the "single server address" and the "nth server address" cases. Your document agrees with this naming convention because you have "setDNS()" vs. "setDns()". But, my API uses an index and your API uses a list of one or two DNS servers. Because I can have more than two DNS servers, I follow this convention:
* `void setDNSServerIP(index, IP)`
* `IPAddress dnsServerIP(index)`
* From Ethernet API: `void setDNSServerIP(IP)`
* From Ethernet API: `IPAddress dnsServerIP()`

I don't want to use the [`setDNS(a, b)`](https://www.arduino.cc/reference/en/libraries/wifinina/wifi.setdns/) form because it's inconsistent with the `dnsIP(n)` form (see your PR [here](https://github.com/arduino-libraries/WiFiNINA/pull/251)). Both should use an index. This is why I have `setDNSServerIP(index, IP)` and `dnsServerIP(index)`.

In summary, the correct way in _QNEthernet_ to get the "nth" DNS server IP is `dnsServerIP(n)` and the correct way to set it is `setDNSServerIP(n, IP)`. Does that clarify things?

**Comment by ssilverman** [comment]
I'll add a few miscellaneous notes I was just thinking about:
1. The name "setDNS" isn't symmetrical with "dnsIP".
2. All the other IP-related functions in the Arduino Ethernet reference are suffixed with the word "IP". (But I don't love how "setDNSIP(x)" looks.)
3. I need to keep "setDnsServerIP", at least as an available option, because some code might use it, since it's the original "Arduino Ethernet API way".

**Comment by JAndrassy** [comment]
Sorry, yes, I would recommend Ethernet libraries to have function names like the legacy Ethernet library. I call it legacy Ethernet API because of the trend of WiFi API winning. Many LwIP based Arduino libraries support both WiFi and Ethernet API usually for MCU peripherals or specific board peripherals. This causes mixing functions from both API in a common class for a network interface. There are more WiFi libraries and the WiFi API has more methods so I expect the WiFi API to slowly take over. Basically a `new` or `modern` Ethernet API evolving. But I would still recommend to have `Legacy Ethernet API` functions. 

In the API usage overview tables I use WiFi API names in the title (if the WiFi API has that function) just because there are more WiFi libraries and some Ethernet libraries have the WiFi version too. Maybe I should make some flag for the legacy Ethernet API cells.

APIs evolving over a long time sometimes in separate libraries and the combined back are usually inconsistent in many ways.

It is your library so you can have any functions and function names you want, of course. I only recommend to have the set of functions of the legacy Ethernet API. And maybe having the WiFi alternatives for users coming from WiFi libraries.

There are more documents in the repo now, all still work in progress.  I still continue the research in details, I wrote test sketches to compare compilation and working of the libraries and I have discussions like this with developers in different repositories. 

You could add WiFi support to your library too :-)

**Comment by JAndrassy** [comment]
to `setDNS` name. some think the S stands for 'server' so they think it is like "LCD disc" to have "DNSServer". (but It is domain name system server)

and you can't set/add DNS server other way than with IP. so why have "IP" in the name?

**Comment by ssilverman** [comment]
I think my main point is that in the table, the most correct thing to put for my library is “setDNSServerIP” vs. “setDnsServerIP”, for now. (We can always adjust in the future, as needed. This is just my request.) I’ll do some thinking about supporting the others, I just don’t love how “setDNS” and “dnsIP” have asymmetrical names and also asymmetrical arguments types and behaviours.

Thanks again for caring about standards and standardization.

**Comment by JAndrassy** [comment]
`setDNS(ip1, ip2)` was already in the first WiFi library by Arduino and dnsIP(n) originates from ESP82666WiFi library where it was added next to localIP() and gatewayIP()

the only way to set/add a DNS server is IP address. but you can query more than IP about a DNS server.

**Comment by JAndrassy** [comment]
I edited the "[Network interface getters and setters](https://github.com/JAndrassy/Arduino-Networking-API/blob/main/ArduinoNetAPILibs.md#network-interface-getters-and-setters)"

**Comment by ssilverman** [comment]
Cool. I'm going to think about it some more.

**Comment by ssilverman** [comment]
Closing for now. I’ll come back with another issue if I have more requests.
